api/src/main/java/io/minio/ChunkedInputStream.java

```diff
@@ -0,0 +1,205 @@
+    }
+
+    int pos = 0;
+    int len = buf.length;
+    int totalBytesRead = 0;
+    int bytesRead = 0;
+    while (totalBytesRead < buf.length) {
+      bytesRead = inputStream.read(buf, pos, len);
+      if (bytesRead < 0) {
+        this.isEof = true;
+        break;
+      }
+
+      totalBytesRead += bytesRead;
+      pos = bytesRead;
+      len = buf.length - totalBytesRead;
+    }
+
+    return totalBytesRead;
+  }
+
+
+  private void createChunkBody(byte[] chunk)
+    throws IOException, NoSuchAlgorithmException, InvalidKeyException, InsufficientDataException, InternalException {
+    String chunkSha256 = Digest.sha256Hash(chunk, chunk.length);
+    String signature = Signer.getChunkSignature(chunkSha256, this.date, this.region, this.secretKey,
+                                                this.prevSignature);
+
+    ByteArrayOutputStream os = new ByteArrayOutputStream();
+    // Add metadata.
+    os.write(Integer.toHexString(chunk.length).getBytes(StandardCharsets.UTF_8));
+    os.write(";chunk-signature=".getBytes(StandardCharsets.UTF_8));
+    os.write(signature.getBytes(StandardCharsets.UTF_8));
+    os.write("\r\n".getBytes(StandardCharsets.UTF_8));
+    // Add chunk data.
+    os.write(chunk, 0, chunk.length);
+    os.write("\r\n".getBytes(StandardCharsets.UTF_8));
+
+    this.chunkBody = os.toByteArray();
+    this.chunkPos = 0;
+    this.prevSignature = signature;
+  }
+
+
+  private int readChunk(int chunkSize)
+    throws IOException, NoSuchAlgorithmException, InvalidKeyException, InsufficientDataException, InternalException {
+    byte[] chunk = new byte[chunkSize];
+    int len = readData(chunk);
+    if (len < 0) {
+      return -1;
+    }
+
+    if (len != chunkSize) {
+      throw new InsufficientDataException("Insufficient data.  read = " + len + " expected = " + chunkSize);
+    }
+
+    createChunkBody(chunk);
+    return this.chunkBody.length;
+  }
+
+
+  /**
+   * read single byte from chunk body.
+   */
+  public int read() throws IOException {
+    if (this.bytesRead == this.length) {
+      // All chunks and final additional chunk are read.
+      // This means we have reached EOF.
+      return -1;
+    }
+
+    try {
+      if (this.streamBytesRead == 0 || this.chunkPos == this.chunkBody.length) {
+        if (this.streamBytesRead != this.streamSize) {
+          // Send all chunks.
+          int chunkSize = CHUNK_SIZE;
+          if (this.streamBytesRead + chunkSize > this.streamSize) {
+            chunkSize = this.streamSize - this.streamBytesRead;
+          }
+
+          if (readChunk(chunkSize) < 0) {
+            return -1;
+          }
+
+          this.streamBytesRead += chunkSize;
+        } else {
+          // Send final additional chunk to complete chunk upload.
+          byte[] chunk = new byte[0];
+          createChunkBody(chunk);
+        }
+      }
+
+      this.bytesRead++;
+      // Return value must be between 0 to 255.
+      return this.chunkBody[this.chunkPos++] & 0xFF;
```

Comment:
How does `this.chunkBody[this.chunkPos++]` evaluate? Does it return the value at index `this.chunkPos` before the increment? Is it non-idiomatic to do the increment and indexing in separate statements?